src/decoder.rs

```diff
@@ -75,24 +80,284 @@ impl Decoder {
+                                next_inst.rd(),
+                            );
+                            let fuze_size = head_size + next_size;
+                            Ok(set_instruction_length_n(fuze_inst.0, fuze_size))
+                        } else {
+                            Ok(head_instruction)
+                        }
+                    }
+                    _ => Ok(head_instruction),
+                }
+            }
+            insts::OP_DIVU => {
+                let head_inst = Rtype(head_instruction);
+                let head_size = instruction_length(head_instruction);
+                let next_instruction_opt = self.decode_raw(memory, pc + head_size as u64);
+                if next_instruction_opt.is_err() {
+                    return Ok(head_instruction);
+                }
+                let next_instruction = next_instruction_opt.unwrap();
+                let next_opcode = extract_opcode(next_instruction);
+                match next_opcode {
+                    insts::OP_REMU => {
+                        let next_inst = Rtype(next_instruction);
+                        if head_inst.rd() != head_inst.rs1()
+                            && head_inst.rd() != head_inst.rs2()
+                            && head_inst.rs1() == next_inst.rs1()
+                            && head_inst.rs2() == next_inst.rs2()
+                        {
+                            let next_size = instruction_length(next_instruction);
+                            let fuze_inst = R4type::new(
+                                insts::OP_WIDE_DIVU,
+                                head_inst.rd(),
+                                head_inst.rs1(),
+                                head_inst.rs2(),
+                                next_inst.rd(),
+                            );
+                            let fuze_size = head_size + next_size;
+                            Ok(set_instruction_length_n(fuze_inst.0, fuze_size))
+                        } else {
+                            Ok(head_instruction)
+                        }
+                    }
+                    _ => Ok(head_instruction),
+                }
+            }
+            _ => Ok(head_instruction),
+        }
+    }
+
+    pub fn decode<M: Memory>(&self, memory: &mut M, pc: u64) -> Result<Instruction, Error> {
+        if self.mop {
+            self.decode_mop(memory, pc)
+        } else {
+            self.decode_raw(memory, pc)
+        }
+    }
 }
 
-pub fn build_decoder<R: Register>(isa: u8, version: u32) -> Decoder {
-    let mut decoder = Decoder::new(version);
+pub fn build_decoder<R: Register>(isa: u8) -> Decoder {
+    let mut decoder = Decoder::new();
     decoder.add_instruction_factory(rvc::factory::<R>);
     decoder.add_instruction_factory(i::factory::<R>);
     decoder.add_instruction_factory(m::factory::<R>);
     if isa & ISA_B != 0 {
         decoder.add_instruction_factory(b::factory::<R>);
     }
+    decoder.mop = isa & ISA_MOP != 0;
```

Comment:
This feels weird to me, why not make `mop` a parameter of `Decoder::new`?

Reply:
It's a design defect, fixed now.